contribute/design/1.x/ndm/20190905-ndm-exporter-integration.md

```diff
@@ -0,0 +1,180 @@
+you need to fetch from the cluster.
+
+##### Cluster level Exporter
+This component of the exporter will run at the cluster level and there will be only one
+running instance of this component at any time in the cluster. The primary responsibility
+of this exporter is to take static / rarely changing data about the block devices from
+etcd and expose it as prometheus metrics. The static data like Model number, UUID etc
+will be cached in the exporter and metrics like state(Offline/Online) of the device will be
+fetched from etcd everytime a request comes at the end point for the metrics
+
+##### Node level Exporter
+The node level exporter runs as a daemon in each node and collects all the metrics of the
+block device like IOPS, temperature, free space etc. It exposes a rest endpoint which can
+be queried to get the data. Everytime the exporter is queried, all the metrics related are
+fetched from the disk using SMART and Seachest libraries and returned.
+
+##### Workflow
+```
++---------------+                      +---------------+      2      +---------+
+|               |         1            |               +------------>+         |
+|               +--------------------->+ cluster level |             | etcd    |
+|   Prometheus  |                      | exporter      +<------------+         |
+|               |                      |               |      3      +---------+
+|               +<---------------------+               |
+|               |         4            |               |
+++--+---+----+--+                      +---------------+
+ |  ^   |    ^
+ |  |   |    +-----------------8-------------------------+
+ |  |   |                                                |
+ |  |   +-----------------5----------------------+       |
+ 5  |                                            |       |
+ |  8           +----------------+            +--v-------+-----+
+ |  |           |                |            |                |
+ |  +-----------+                |            |                |
+ |              |      Node1     |            |      Node2     |
+ |              |                |            |                |
+ +------------->+ exporter-d     |            |  exporter-d    |
+                +--------+-------+            +--------+-------+
+                     6|  ^                        6|   ^
+                      v  |7                        v   |7
+                    +-+-------+                  +-+-------+
+                    |::Disk   |                  |::Disk   |
+                    +---------+                  +---------+
+
+```
+
+- Cluster Level exporter
+  1. Queries the cluster level operator for static data about all block devices
+     in the cluster
+  2. Requests etcd for all blockdevices and their properties
+  3. Response from etcd with all blockdevices
+  4. Response from cluster level exporter to prometheus with static metrics
+
+- Node Level exporter
```

Comment:
There are two exporters, we need both the information combined to get the complete picture?  can I see all the data for a device (uid) from both the exporters together in prometheus?

Reply:
Yes @pawanpraka1 . It can be seen. Data from multiple exporters can be joined together using the labels in the metrics and useful information can be extracted out of it. Also metrics from Prometheus node-exporter can also be used along with the metrics given by NDM exporter.